pkg/kubelet/container/runtime.go

```diff
@@ -282,6 +284,8 @@ type Container struct {
 	Image string
 	// The id of the image used by the container.
 	ImageID string
+	// Runtime handler used to pull the image if any.
+	ImageRuntimeHandler string
```

Comment:
I get why `ImageSpec` was updated to be inclusive of `RuntimeHandler`, and how that is convenient when runtime manager `ListImages` is invoked (where there is no associated pod or container context).

What is not clear to me is why `Container` is extended to have `ImageRuntimeHandler` rather than `Pod`.  The `RuntimeClassName` is a pod level attribute, and it looks like every other call site where we work with `Container` we have access to the encapsulating `Pod` context.  Worried in future we could have a mismapped reference of some kind.  I do not think this is blocking on the PR, but it felt more natural to me to model that runtime handler against the pod since the API facing the users has the field at `Pod` scope.

Reply:
![image](https://github.com/kubernetes/kubernetes/assets/99994218/c71ab1b7-e6c4-4c7e-aacd-2123daef9180)

I had to do this because of how image GC is reading imageID (its doing it from the kuberuntime.Container.ImageId) and the image is referred to by just the imageID in kuberuntime.Container which should not be the case once we move to identifying images as (imageID, runtimeHandler).

Reply:
If we do not want to add runtimeHandler to kuberuntime.Container, I think we should get rid of the imageId too as it might cause issues in the future if we just refer to the image that a container is using by the imageID and not (imageId, runtimeHandler). We could have the same ImageID for different runtime handlers and it would be risky to just reference the image by ImageID only

Reply:
@kiashok that was the code block that was tripping me up.  `ListImages` needs to include the runtime handler, totally agree.  If runtimeHandler was added to `Pod`, it would be returned in `runtime.GetPods` with that pod.

In the code block you capture above, would imagesInUse would just use the imageKey of `container.ImageID,pod.RuntimeHandler`, no?

Reply:
reflecting a bit on the comment to remove `imageID`, i guess that is fair.  The issue is that is the ID according to the runtime, its just only identifying when now associated with a runtime handler... I think that is a separate issue we should think about how to cleanup, but for this particular instance, it feels like Pod is probably the right place to associate runtime handler information in kuberuntime.

Reply:
Hi @derekwaynecarr , I think Container struct having the imageID is kinda expected since container runtimes also use imageID to identify which image the container is actively using now. This also seems to be expected from Container struct in CRI:

![image](https://github.com/kubernetes/kubernetes/assets/99994218/bccc0cb3-4190-41ce-a8a1-4332d2d97ec7)

It would be useful to know the imageID of the image that the container is currently using right? I am not sure we would want to get rid of it completely though as container runtimes might need to know what image a container needs and if its present or not right? I think now that we are changing how an image is identified, we would need to just have info about runtimeHandler as well. That's why I named the field as "ImageRuntimeHandler" with that comment that it is the runtimehandler used to pull the image. I still feel its important to identify what image is being used by container. Any thoughts?

Reply:
> In the code block you capture above, would we not just insert into imagesInUse the tuple key composed of container.ImageID and pod.RuntimeHandler?

Initially makes sense to me


@kiashok  If I understand correctly this concern is there might be a mis match between what containerd knows about the image and what kubelet does? 

One thing I noticed is the cri `container` message shouldn't need `ImageRuntimeHandler` since it already has the `Image` reference in the `container` message and `ImageSpec` will have that information? https://github.com/kubernetes/kubernetes/blob/029452198566a41bc39d04a1ec5bad3f37621a1c/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L1195

Would that resolve the concern of a mis-match and not require that field added?

Reply:
from a container... should be able to find the runtime handler like so: 
pod := podSandboxMap[container.podSandboxID]   m.runtimeClassManager.LookupRuntimeHandler(pod.Spec.RuntimeClassName)

Reply:
@jsturtevant , I was referring to kubecontainer.Container struct and not CRI container. CRI container struct already has reference to ImageSpec which is correct. kubecontainer.Container struct in pkg/kubelet/container/runtime.go has its own struct and saves info as needed:

![image](https://github.com/kubernetes/kubernetes/assets/99994218/3e4e73ba-e2da-4c3c-8e78-dab25a8cb5de)


I think the concern is we do not need to have imageRuntimeHandler in this struct.

I think it would be good to have runtimehandler here as they are using imageID as reference to the image. It might cause confusions as image is no longer just recognized by the imageID

Reply:
@mikebrow I know we could go backwards and get it. My concern is that we should make image runtime handler explicit in structures that are using imageID to reference the image currently in use by the container

Reply:
If container CRI has reference to ImageSpec which has a field for runtimeHandler, we should maintain uniformity and also have image runtime handler with kuberuntime.Container?

Reply:
> @kiashok that was the code block that was tripping me up. `ListImages` needs to include the runtime handler, totally agree. If runtimeHandler was added to `Pod`, it would be returned in `runtime.GetPods` with that pod.
> 
> In the code block you capture above, would we not just insert into `imagesInUse` the tuple key composed of `container.ImageID` and `pod.RuntimeHandler`?

@derekwaynecarr , that code is changed in the current PR. I was just pointing out to what it is like in the current kubelet package.

for reference:

Reply:
Since this is not a blocker , can this be added as a follow up in beta after further discussion?

Reply:
I’m good with follow up.

Reply:
nod.. pls add a TODO for follow up revision

Reply:
Discussed with @mikebrow  on slack. This will be TODO for beta and I'll add this to sig-node meeting agenda for the next meeting and have a wider discussion with folks before updating the repo with these changes. thanks!